(PDB-1674) PuppetDB hangs on shutdown when the database is down

senior opened the pull request:
This adds connection timeouts to the command processing and query
connection pools. This fixes the problem where the ActiveMQ session
can't be properly shutdown because the command connection pool can't get
a connection to the database (i.e. the database is down). Previously
this would hang, even with a Control-C. This hang was due to the
connection pool software getting caught in a loop trying to get a
connection to the database.

This connection timeout was also applied to the query side of the
application as user provided queries are expected to respond quickly and
a hanging behavior would be bad there as well.

This blocking/retrying behavior is what we want when PuppetDB first
starts up as we need to initialize the schema etc. This commit also
creates a special connection pool (just for startup) that doesn't have
these timeouts. This pool is closed after the startup code that needs a
database connection is finished.

senior commented:
This just needs docs for the new connection-timeout config

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1573/

senior commented:
@pljenkinsro retest this please

ajroetker commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -237,6 +237,35 @@
+  be opened and closed within the body of this function."
+  [write-db-config product-name]
+  (with-open [init-db-pool (pl-jdbc/make-connection-pool (assoc write-db-config
+                                                           ;; Block waiting to grab a connection
```
this alignment looks off

senior replied:
Maybe the comment is what throws it off? I reformatted it and nothing changed.

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1575/

senior commented:
Added some docs, switched from 2ms connection timeout on the query side to 500ms. Travis is failing with connection timeouts, only on PG. I'm wondering if the issue it's having is lazy init. The first poll for a database connection needs to init the pool, which maybe takes more than 2ms?

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1592/

senior commented:
1 transient failing cell ^^^

rbrw commented:
Not sure which test is failing, but I also wouldn't be surprised if some tests might need to be adjusted to account for these changes, i.e. a query EAGAIN is now more "normal".

rbrw commented on documentation/configure.markdown:
```diff
@@ -370,6 +370,10 @@ The maximum time (in minutes) a pooled connection should remain open. Any connec
+
+The maximum time to wait (in milliseconds) to acquire a connection from the pool of database connections. If not supplied, defaults to 1000.
+
 ###`statements-cache-size`
```
Perhaps wrap this (and below)?

senior replied:
Meh, I do like the wrap there, but the stuff around it is not wrapped. The whole file goes in and out of it, but specifically those sections are not. Probably something we should just do on all files at once if we want the change.

rbrw commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -237,6 +237,35 @@
+  creating anything."
+  [db-conn-pool product-name]
+  (sql/with-connection db-conn-pool
+    (scf-store/validate-database-version #(System/exit 1))
```
Be glad when we finally get a chance to consider our System/exits (but not an issue for this PR).

rbrw commented:
This all looks good to me, works fine, and I liked the commit breakdown too.
